Landing Page - What is WheelNext (#8)
wheelnext/wheelnext: docs/index.md

# Welcome to WheelNext

## What is WheelNext and Why should you participate !?

WheelNext is an open-source initiative (<https://github.com/wheelnext> & <https://wheelnext.dev/>) aiming to improve the

> user-experience in the Python packaging ecosystem, specifically around the scientific computing and machine/deep learning space.

Reviewer: ```suggestion
user experience in the Python packaging ecosystem, specifically around the scientific computing and machine/deep learning space.
```

Reviewer: I'd add something like "We also anticipate benefits in other domains that heavily rely on performance of compiled Python extension modules - the benefit of utilizing one's hardware more optimally is not exclusive to any single domain."

Author: Suggestion added and merged

> The current state of Python packaging was designed and implemented at a time before the increasingly complex needs these use-cases demand were prevalent.

Reviewer: ```suggestion
The current state of Python packaging was designed and implemented at a time before the increasingly complex needs these use cases demand were prevalent.
```

> Through WheelNext, we aim to evolve Python packaging standards to better address the needs of this important segement, while retaining its effectiveness for the overall ecosystem.

Reviewer: ```suggestion
Through WheelNext, we aim to evolve Python packaging standards to better address the needs of this important segment, while retaining its effectiveness for the overall ecosystem.
```

Many of the problems and gaps are summarized on the following website: <https://pypackaging-native.github.io/>
Objective: To improve the state of packaging for Python native code, making it easier to build, distribute, and install
Python packages with complex native dependencies.

> ### Key Issues Identified:

Reviewer: It looks to me like @warsaw's comments on some of the bullets are on point, and can be made about the other bullet points as well. E.g., we're probably going to introduce more rather than less Complexity.

It's probably best to be more concrete. The key issues are lack of support for:
- GPUs and other accelerators
- CPU capabilities beyond the CPU family (SIMD instruction sets)
- dependencies meant to be used as a single runtime or shared library rather than vendored (e.g., OpenMP, MPI, BLAS, LAPACK)
- paper cuts specific to packaging complex native libraries as wheels (e.g., symlink support)

Author: @rgommers do you mind making suggestions to the PR ? Much easier to capture a phrasing that fairly represent your thinking if it comes directly from you

Reviewer: Sure, if you're happy for me to rewrite this whole section. I'll try now, and point you at a commit since inline suggestions won't work for such a large block.

Reviewer: Here it is: https://github.com/rgommers/wheelnext/commit/ebc4e6faf39a365b385670a2de1ccdf6c83b2740

It could be improved further, but I think it's already in reasonable shape. A key point is to remove the Complexity, Tooling Fragmentation, Lack of Standardization, Dependency Management items, because some of those can be argued with and none of those are things WheelNext is going to solve.

Author: Perfect. Thanks it was merged

- **Complexity:** Building and packaging native code in Python is complex and often inconsistent across different platforms.
- **Tooling Fragmentation:** Multiple tools and standards exist, leading to confusion and inefficiency.
- **Dependency Management:** Managing dependencies, especially for native code, is challenging.
- **Installation Issues:** Users often face difficulties installing packages with native code due to environment
specific issues.

> - **Lack of Standardization:** There is no single, standardized way to handle native packaging in Python.

Reviewer 2: Is this true?  I mean, there *is* a standard, but it's not sufficient for the complexity WN is going after.

Author: Taken from that end - true.
I was sort of thinking of out-of-standard strategies people have come up with to address this. Like for example `conda` or `uv` in one recent PR to inject the right index for pytorch

> - **No Package Index Prioritization:** The main package installer `pip` is lacking ability to prioritize a given package

Reviewer 2: What's the more generic problem that this bullet point covers?

Author: I'm rephrasing this - It's not good

over another.
- **Lack of Symlink Support:** Very important to reduce package and better support native code. Yet it is lacking in the
packaging standard.

> - **No ability to build a binary for a specific architecture:** We are lacking the ability to describe which platform or

Reviewer 2: ```suggestion
- **No ability to build a binary for a specific microarchitecture:** We are lacking the ability to describe with specificity the platform,
```

> architecture is supported by a given binary (ARMv7, ARMv8, etc.)

Reviewer 2: ```suggestion
microarchitecture, or other execution environment details is supported by a given binary (ARMv7, ARMv8, etc.)
```

> - **No clear path to support non-default package indexes:** Using private / corporate package indexes is common

Reviewer 2: Actually, the tooling does have ways to specify which indexes other than the default, a user wants to use, it's just that it's very difficult to safely and precisely specify how multiply enabled indices interoperate to resolve the requested top-level and transitive dependencies.

Author: Fair - rephrasing trying to re-use your wording :)

practice, yet the tooling ecosystem does not have a solid and reliable way to support these workflows.

> ### Proposed Solutions:

Reviewer: I suggest renaming to "Approach to Proposed Solutions", because they're more about the philosophy or abstract approach rather than concrete solutions.

Author: Agreed & Done

- **Unified Approach:** Developing a more unified and standardized approach to packaging native code in Python.
- **Improved Tooling:** Enhancing existing tools and creating new ones to simplify the process.
- **Better Documentation:** Providing comprehensive and clear documentation to help developers navigate the packaging landscape.

> - **Engineering Push:** Putting the work and the effort to build and propose clear solutions that fixes the

Reviewer 2: ```suggestion
- **Engineering Focus:** Putting the work and the effort to build and propose clear solutions that fixes the
```

- Proposing new topics for inclusion on the website, and writing the content for them
- ... and anything else you consider useful!

The content for this website is [maintained on GitHub](https://github.com/wheelnext/wheelnext).

## Acknowledgements

> - Initial kickoff effort was led by [NVIDIA](http://nvidia.com/), now aggregating efforts from many companies.

Reviewer: ```suggestion
- Initial kickoff effort was led by [NVIDIA](http://nvidia.com/), now aggregating efforts from many companies,
 maintainers of open source projects, and other contributors.
```
